lcls-twincat-optics/lcls_twincat_optics_plc/POUs/Helpers/FB_MirrorTwoCoatingProtection.TcPOU

```diff
@@ -3,57 +3,110 @@
-
-	nUpperCoatingBoundary : UDINT; // Encoder count for upper boundary
-	nUpperCoatingBitmask : DWORD; // Bitmask for upper coating
-	sUpperCoatingType : STRING := ''; // Type of coating
-	
-	nLowerCoatingBoundary : UDINT; // Encoder count for lower boundary
-	nLowerCoatingBitMask : DWORD;
-	sLowerCoatingType : STRING := ''; // Type of coating
-    
+    nCurrentEncoderCount : UDINT; // Current encoder count
+    neVRange : DWORD; // Current ev range from stCurrentBeamParams
+
+    sDevName : STRING := ''; // Device name
+
+    nUpperCoatingBoundary : UDINT; // Encoder count for upper boundary
+
+    sUpperCoatingType : STRING := ''; // Type of coating
+
+    nLowerCoatingBoundary : UDINT; // Encoder count for lower boundary
+
+    sLowerCoatingType : STRING := ''; // Type of coating
+
     bAutoClear : BOOL := TRUE; // Auto-clear these fast faults
+
+    bReadPmpsDb : BOOL; // Trigger a re-read of the JSON Beam Parameters
 END_VAR
 VAR_OUTPUT
 END_VAR
 VAR_IN_OUT
-	FFO : FB_HardwareFFOutput;
+    FFO : FB_HardwareFFOutput;
 END_VAR
 VAR
-	ffUpperCoating: FB_FastFault := (
-		i_xAutoReset := FALSE,
-		i_TypeCode := 16#401);
-	ffLowerCoating : FB_FastFault := (
-		i_xAutoReset := FALSE,
-		i_TypeCode := 16#401);
-		
-	bInit : BOOL;
+    ffUpperCoating: FB_FastFault := (
+        i_xAutoReset := FALSE,
+        i_TypeCode := 16#401);
+    ffLowerCoating : FB_FastFault := (
+        i_xAutoReset := FALSE,
+        i_TypeCode := 16#401);
+    sFFBeamParamMismatch : STRING := ' mirror coating incompatible with beam photon energy';
+    sFFBeamParamNotLoaded : STRING := ' mirror coating beam parameters not loaded';
+
+    aDbStateParams : ARRAY[0..1] OF ST_DbStateParams;
+    fbGetCoatingBPs : FB_JsonDocToSafeBP;
+    ftReadJsonDocDone : F_TRIG;
+    bBPsLoaded : BOOL := FALSE;
+
+    nUpperCoatingBitmask : DWORD := 0;
+    nLowerCoatingBitMask : DWORD := 0;
+
+    i : INT;
+    sDevState : STRING  := '';
+    bInit : BOOL;
 END_VAR
 ]]></Declaration>
     <Implementation>
       <ST><![CDATA[IF NOT bInit THEN
-	ffUpperCoating.i_Desc := CONCAT(sUpperCoatingType, ' mirror coating incompatible with beam photon energy');
-	ffUpperCoating.i_DevName := sDevName;
-	
-	ffLowerCoating.i_Desc := CONCAT(sLowerCoatingType, ' mirror coating incompatible with beam photon energy');
-	ffLowerCoating.i_DevName := sDevName;
-	
-	bInit := TRUE;
+    ffUpperCoating.i_DevName := sDevName;
+    ffLowerCoating.i_DevName := sDevName;
+
+    FOR i:=0 to 1 BY 1 DO
+        sDevState := CONCAT(sDevState, sDevName);
+        sDevState := CONCAT(sDevState, '-');
+        CASE i OF
+        0:
+            sDevState := CONCAT(sDevState, sUpperCoatingType);
+        1:
+            sDevState := CONCAT(sDevState, sLowerCoatingType);
+        END_CASE
+        aDbStateParams[i].sPmpsState := sDevState;
+        sDevState := '';
+    END_FOR
+    bInit := TRUE;
+END_IF
+
+IF bReadPmpsDB THEN
+    bBPsLoaded := FALSE;
+END_IF
+
+fbGetCoatingBPs(bExecute := bReadPmpsDB,
```

Comment:
This should work pretty well. The FB in use here is only looking for rising edges and is ignoring the falling edges. Some longer-running FBs will turn themselves off on falling edges but this one is intended to be pretty quick so it doesn't bother.